components/lib/state/src/main/java/mozilla/components/lib/state/helpers/Helpers.kt

```diff
@@ -8,34 +8,43 @@ import mozilla.components.lib.state.Observer
 import mozilla.components.lib.state.State
 
 /**
- * Creates an [Observer] that will map the received [State] to [T] (using [map]) and will invoke the callback
- * [then] only if the value has changed from the last mapped value.
+ * Creates an [Observer] that will map the received [State] to [T] (using [map]) and will invoke the
+ * callback [then] only if the value has changed from the last mapped value. If rhe mapped value
+ * changed to `null` then the optional [otherwise] callback gets invoked.
  *
  * @param onMainThread Whether or not the [then] function should be executed on the main thread.
  * @param map A function that maps [State] to the value we want to observe for changes.
  * @param then Function that gets invoked whenever the mapped value changed.
+ * @param otherwise Optional function that gets invoked whenever the mapped value is `null`.
  */
 @Suppress("FunctionNaming")
 fun <S : State, T> onlyIfChanged(
```

Comment:
I feel like this function is getting too complex. We should have separate `onlyIfChanged` and `map` functions that can be composed.

Reply:
I agree. @csadilek and I wanted to come back to this function at some point and think of better approaches. We haven't filed anything for that though - so I did that now:
https://github.com/mozilla-mobile/android-components/issues/3786